File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res=0, tmp=0, p1=0, p2=0, l=s.length();
        unordered_map<char,int>m;
        
        while(p2 < l){
            m[s[p2]]++;
            
            if(m[s[p2]] > 1){
                p1++;
                p2=p1;
                m.clear();
                if(tmp > res) res=tmp;
                tmp=0;
            }
            else {
                tmp++;
                p2++;
            }
        }
        if(tmp > res) res=tmp;
        
        return res;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s)==0: return 0
        
        m=defaultdict(lambda:-1)
        
        length=0
        lp, rp=0, 0
        
        while rp<len(s):
            if m[s[rp]]>=0 and m[s[rp]]>=lp:
                if rp-lp>length:
                    length=rp-lp
                
                lp=m[s[rp]]+1
            
            m[s[rp]]=rp
            rp+=1
        
        if rp-lp>length:
            length=rp-lp
        
        return length